File: 2020/python/03/03.py
def inputParser(path:str) -> list:
    lines = []
    with open(path) as f:
        for l in f:
            lines.append(l[:-2])
    return lines



def main():
    levels =  inputParser("input2.txt") # Matrix
    lv = 0
    i = 0
    cant_levels = len(levels)
    largo_lv = len(levels[0]) 
    print(levels[0]) 
    print(f"{cant_levels} - {largo_lv}")
    trees = 0
    while True:
        print(f"{lv} - {i}")
        print(levels[lv])
        lv= lv + 1
        if  lv >= cant_levels:
            return trees
        i += 3
        if i >= largo_lv:
            i = i - largo_lv
        if levels[lv][i] == "#":
            trees = trees + 1
            print("#")

if __name__ == "__main__" :
    print(main())

File: 2020/go/03/03.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func fileToSlice(fp string) ([]string, error) {
	var lines []string
	f, err := os.Open(fp)
	check(err)
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	defer f.Close()
	return lines, scanner.Err()
}

func count_trees(path []string, du int, rl int) int {
	l1 := len(path[0])
	l2 := len(path)
	fmt.Printf("%d - %d\n", l1, l2)
	j := 0
	i := 0
	trees := 0
	for {
		fmt.Printf("%d - %d\n", j, i)
		fmt.Printf("%v\n ", path[j])
		j = j + du
		if j >= l2 {
			return trees
		}
		i = i + rl
		if i >= l1 {
			i = i - l1
		}
		if string(path[j][i]) == "#" {
			trees++
			fmt.Println("#")
		}
	}
}

func main() {
	path, err := fileToSlice("example.txt")
	//path, err := fileToSlice("input.txt")
	check(err)
	trees := count_trees(path, 1, 3)
	fmt.Printf("Number of trees in the path: %d \n", trees)
	slopes := [][]int{{1, 1}, {5, 1}, {7, 1}, {1, 2}}
	for _, s := range slopes {
		trees *= count_trees(path, s[1], s[0])
	}
	fmt.Printf("Number of trees in all paths: %d \n", trees)
}
